### PR TITLE
build: promote tsan ci to must-pass

### PR DESCRIPTION
**Pull request by bnoordhuis**
Refs: https://github.com/libuv/libuv/issues/3681

<hr>

Not completely ready, see discussion in the linked issue.

### CODE REVIEW

**Comment by vtjnash** [comment]
LGTM

**Comment by bnoordhuis** [comment]
The `barrier_serial_thread` indeed fails, as expected:
<details>

```
not ok 9 - barrier_serial_thread
# exit code 66
# Output from process `barrier_serial_thread`:
# ==================
# WARNING: ThreadSanitizer: data race (pid=10538)
#   Write of size 1 at 0x7ffe3fbd37b0 by main thread:
#     #0 pthread_barrier_destroy ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1413 (libtsan.so.0+0x35d06)
#     #1 uv_barrier_destroy <null> (uv_run_tests_a+0xcba41)
#     #2 run_test_barrier_serial_thread <null> (uv_run_tests_a+0x171f0)
#     #3 run_test_part <null> (uv_run_tests_a+0x159a2)
#     #4 main <null> (uv_run_tests_a+0x105a1)
# 
#   Previous read of size 1 at 0x7ffe3fbd37b0 by thread T4:
#     #0 pthread_barrier_wait ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1420 (libtsan.so.0+0x4e0e8)
#     #1 uv_barrier_wait <null> (uv_run_tests_a+0xcb9fe)
#     #2 serial_worker <null> (uv_run_tests_a+0x16ca1)
# 
#   Location is stack of main thread.
# 
#   Location is global '<null>' at 0x000000000000 ([stack]+0x00000001f7b0)
# 
#   Thread T4 (tid=10545, running) created by main thread at:
#     #0 pthread_create ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:969 (libtsan.so.0+
# 0x605b8)
#     #1 uv_thread_create <null> (uv_run_tests_a+0xcbc2c)
#     #2 run_test_barrier_serial_thread <null> (uv_run_tests_a+0x171cf)
#     #3 run_test_part <null> (uv_run_tests_a+0x159a2)
#     #4 main <null> (uv_run_tests_a+0x105a1)
# 
# SUMMARY: ThreadSanitizer: data race (/home/runner/work/libuv/libuv/build-tsan/uv_run_tests_a+0xcba41) in uv_barrier_destroy
# ==================
```

</details>

I've added a suppression file.

File: .github/workflows/sanitizer.yml [hunk 1]
```diff
@@ -47,9 +47,8 @@ jobs:
           (cd build-tsan && cmake .. -G Ninja -DBUILD_TESTING=ON -DTSAN=ON -DCMAKE_BUILD_TYPE=Release)
           cmake --build build-tsan
       - name: TSAN Test
-        continue-on-error: true # currently permit failures
         run: |
-          ./build-tsan/uv_run_tests_a
+          env TSAN_OPTIONS=tsansupp.txt ./build-tsan/uv_run_tests_a
```
Comment (vtjnash) [inline .github/workflows/sanitizer.yml:51]:
```suggestion
          env TSAN_OPTIONS= suppressions=tsansupp.txt ./build-tsan/uv_run_tests_a
```

File: .github/workflows/sanitizer.yml [hunk 2]
```diff
@@ -47,9 +47,8 @@ jobs:
           (cd build-tsan && cmake .. -G Ninja -DBUILD_TESTING=ON -DTSAN=ON -DCMAKE_BUILD_TYPE=Release)
           cmake --build build-tsan
       - name: TSAN Test
-        continue-on-error: true # currently permit failures
         run: |
-          ./build-tsan/uv_run_tests_a
+          env TSAN_OPTIONS= suppressions=tsansupp.txt ./build-tsan/uv_run_tests_a
```
Comment (vtjnash) [inline .github/workflows/sanitizer.yml:51]:
```suggestion
          env TSAN_OPTIONS=suppressions=tsansupp.txt ./build-tsan/uv_run_tests_a
```

**Comment by bnoordhuis** [comment]
Test failures:
```
 # exit code 134
# Output from process `spawn_relative_path`:
# exit_cb
# Assertion failed in /home/runner/work/libuv/libuv/test/test-spawn.c on line 68: exit_status == 1
```
And
```
not ok 337 - tcp_writealot
# timeout
# Output from process `tcp_writealot`: (no output)
Output from process `tcp_writealot`: (no output)
```
tcp_writealot isn't really surprising, probably too slow to run under TSan.

spawn_relative_path... I'm going to guess the `UV_TEST_RUNNER_FD` environment variable gets mangled somehow. I'll investigate.

**Pull request merged**